arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Iterable.kt

```diff
@@ -283,153 +283,36 @@ public inline fun <B, C, D, E, F, G, H, I, J, K, L> Iterable<B>.zip(
-public fun <E, A> Iterable<Validated<E, A>>.sequenceValidated(semigroup: Semigroup<E>): Validated<E, List<A>> =
-  sequence(semigroup)
-
-public fun <E, A> Iterable<Validated<E, A>>.sequence(semigroup: Semigroup<E>): Validated<E, List<A>> =
-  traverse(semigroup, ::identity)
-
-@Deprecated("sequenceValidated is being renamed to sequence to simplify the Arrow API", ReplaceWith("sequence()", "arrow.core.sequence"))
-public fun <E, A> Iterable<ValidatedNel<E, A>>.sequenceValidated(): ValidatedNel<E, List<A>> =
-  sequence()
-
-public fun <E, A> Iterable<ValidatedNel<E, A>>.sequence(): ValidatedNel<E, List<A>> =
-  traverse(Semigroup.nonEmptyList(), ::identity)
-
-@Deprecated("traverseOption is being renamed to traverse to simplify the Arrow API", ReplaceWith("traverse(f)", "arrow.core.traverse"))
-public inline fun <A, B> Iterable<A>.traverseOption(f: (A) -> Option<B>): Option<List<B>> =
-  traverse(f)
-
-@OptIn(ExperimentalTypeInference::class)
-@OverloadResolutionByLambdaReturnType
-public inline fun <A, B> Iterable<A>.traverse(f: (A) -> Option<B>): Option<List<B>> {
-  val destination = ArrayList<B>(collectionSizeOrDefault(10))
-  for (item in this) {
-    when (val res = f(item)) {
-      is Some -> destination.add(res.value)
-      is None -> return res
-    }
-  }
-  return destination.some()
+public inline fun <Error, A, B> Iterable<A>.mapAccumulating(
+  semigroup: Semigroup<Error>,
+  transform: (A) -> Either<Error, B>,
+): Either<Error, List<B>> = with(semigroup) {
+  mapAccumulating({ a, b ->
+    a.combine(b)
+  }, transform)
 }
 
-@Deprecated("sequenceOption is being renamed to sequence to simplify the Arrow API", ReplaceWith("sequence()", "arrow.core.sequence"))
-public fun <A> Iterable<Option<A>>.sequenceOption(): Option<List<A>> =
-  sequence()
-
-public fun <A> Iterable<Option<A>>.sequence(): Option<List<A>> =
-  traverse(::identity)
-
-@Deprecated("traverseNullable is being renamed to traverse to simplify the Arrow API", ReplaceWith("traverse(f)", "arrow.core.traverse"))
-public inline fun <A, B> Iterable<A>.traverseNullable(f: (A) -> B?): List<B>? =
-  traverse(f)
-
-@OptIn(ExperimentalTypeInference::class)
-@OverloadResolutionByLambdaReturnType
-public inline fun <A, B> Iterable<A>.traverse(f: (A) -> B?): List<B>? {
-  val acc = mutableListOf<B>()
-  forEach { a ->
-    val res = f(a)
-    if (res != null) {
-      acc.add(res)
-    } else {
-      return res
-    }
-  }
-  return acc.toList()
-}
-
-@Deprecated("sequenceNullable is being renamed to sequence to simplify the Arrow API", ReplaceWith("sequence()", "arrow.core.sequence"))
-public fun <A> Iterable<A?>.sequenceNullable(): List<A>? =
-  sequence()
-
-public fun <A> Iterable<A?>.sequence(): List<A>? =
-  traverse(::identity)
+public inline fun <Error, A, B> Iterable<A>.mapAccumulating(
+  transform: (A) -> Either<NonEmptyList<Error>, B>,
```

Comment:
Hmm, thinking if this should be `EitherNel<Error, B>` or just `Either<Error, B>` 🤔
I think the latter occurs more often, but perhaps we should do both. Or do both signatures conflict with each other?

Reply:
As we discussed in the bi-weekly Arrow meeting, I think something like this makes most sense.

```kotlin
public inline fun <Error, A, B> Iterable<A>.mapAccumulating(
  transform: (A) -> Either<Error, B>
): Either<NonEmptyList<Error>, List<B>> {
  val right = ArrayList<B>(collectionSizeOrDefault(10))
  val left = ArrayList<Error>()
  forEach { a ->
    when(val res = transform(a)) {
      is Either.Left -> left.add(res.value)
      is Either.Right -> right.add(res.value)
    }
  }
  return left.toNonEmptyListOrNull()?.left() ?: right.right()
}
```